### PR TITLE
add post

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE, fig.width = 6, fig.height = 6)
+```
+
+Today, we resume our exploration of group equivariance. This is the third post in the series. The [first](https://blogs.rstudio.com/ai/posts/2023-03-09-group-equivariant-cnn-1/) was a high-level introduction: What this is all about; how equivariance is operationalized; and why it is of relevance to many deep-learning applications. The [second](https://blogs.rstudio.com/ai/posts/2023-03-27-group-equivariant-cnn-2/) sought to concretize the key ideas by developing a group-equivariant CNN from scratch. That being instructive, but too tedious for practical use, today we look at a carefully designed, highly-performant [library](https://github.com/QUVA-Lab/escnn) that hides the technicalities and enables a convenient workflow.
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:28]:
```suggestion
Today, we resume our exploration of group equivariance. This is the third post in the series. The [first](https://blogs.rstudio.com/ai/posts/2023-03-09-group-equivariant-cnn-1/) was a high-level introduction: what this is all about; how equivariance is operationalized; and why it is of relevance to many deep-learning applications. The [second](https://blogs.rstudio.com/ai/posts/2023-03-27-group-equivariant-cnn-2/) sought to concretize the key ideas by developing a group-equivariant CNN from scratch. That being instructive, but too tedious for practical use, today we look at a carefully designed, highly-performant [library](https://github.com/QUVA-Lab/escnn) that hides the technicalities and enables a convenient workflow.
```

Comment (batpigandme) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:28]: Since it's just an item in the list, you don't capitalize it.

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
...gain set the context. In physics, an all-important concept is that of symmetry[^1], a symmetry being present whenever some quantity is being conserved. But we don't even need to look to science. Examples arise in daily life, and -- otherwise why write about it - in the tasks we apply deep learning to. In daily life: Think about speech -- me stating "it is cold", for example. Formally, or denotation-wise, the sentence will have the same meaning now as in five hours. (Connotations, on the other hand, can and will probably be different!). This is a form of translation symmetry, translation in time. In deep learning: Take image classification. For the usual convolutional neural network, a cat in the center of the image is just that, a cat; a cat on the bottom is, too. But one sleeping, comfortably curled like a half-moon "open to the right", will not be "the same" as one in a mirrored position. Of course, we can train the network to treat both as equivalent by providing training images of cats in both positions, but that is not a scaleable approach. Instead, we'd like to make the network aware of these symmetries, so they are automatically preserved throughout the network architecture.
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:30]: > In daily life: Think about speech -- me stating "it is cold", for example.

This is a little off structurally. I think, since you've already mentioned that there are examples from daily life, maybe you could cut that. Otherwise, I'd integrate it into the sentence rather than separating it into a colon-delimited introduction.

Comment (batpigandme) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:30]: > In deep learning: Take image classification.

OK, now I kind of see what you're doing here, but it was a bit hard to follow. I think rather than have this sentence introducing both everyday-life and deep-learning examples, I'd give each example a fuller introduction than, e.g. "In deep learning." That, or separate them into paragraphs.

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+
+[^1]: This is nicely explained in, for example, Jakob Schwichtenberg's [Physics from symmetry](http://physicsfromsymmetry.com/).
+
+## Purpose and scope of this post
+
+Here, I introduce [`escnn`](https://github.com/QUVA-Lab/escnn), a PyTorch extension that implements forms of group equivariance for CNNs operating on the plane or in (3d) space. The library is used in various, amply illustrated research papers; it is appropriately documented; and it comes with introductory notebooks both relating the math and exercising the code. Why, then, not just refer to the [first notebook](https://github.com/QUVA-Lab/escnn/blob/master/examples/introduction.ipynb), and immediately start using it for some experiment?
+
+In fact, this post should -- as quite a few texts I've written -- be regarded as an introduction to an introduction. To me, this topic seems anything but easy, for various reasons. Of course, there's the math. But as so often in machine learning, you don't need to go to great depths to be able to apply an algorithm correctly. So if not the math itself, what generates the difficulty? For me, it's two things.
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:38]: This is a bit hard to follow in terms of what you're setting up for the paragraphs that come after—that there are two things about "this topic" (not totally sure what that is, but I'm also proofing, so maybe that's my bad) that make it difficult, but it's not the math?

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+
+First, to map my understanding of the mathematical concepts to the terminology used in the library, and from there, to correct use and application. Expressed schematically: We have a concept A, which figures (among other concepts) in technical term (or object class) B. What does my understanding of A tell me about how object class B is to be used correctly? More importantly: How do I use it to best attain my goal C? -- This first difficulty I'll address in a very pragmatic way. I'll neither dwell on mathematical details, nor try to establish the links between A, B, and C in detail. Instead, I'll present the characters[^2] in this story by asking what they're good for.
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:40]:
```suggestion
First, to map my understanding of the mathematical concepts to the terminology used in the library, and from there, to correct use and application. Expressed schematically: We have a concept A, which figures (among other concepts) in technical term (or object class) B. What does my understanding of A tell me about how object class B is to be used correctly? More importantly: How do I use it to best attain my goal C? This first difficulty I'll address in a very pragmatic way. I'll neither dwell on mathematical details, nor try to establish the links between A, B, and C in detail. Instead, I'll present the characters[^2] in this story by asking what they're good for.
```

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+
+Second -- and this will be of relevance to just a subset of readers -- the topic of group equivariance, particularly as applied to image processing, is one where visualizations can be of tremendous help. The quaternity[^3] of conceptual explanation, math, code, and visualization can, together, produce an understanding of emergent-seeming quality... if, and only if, all of these explanation modes "work" for you. (Or if, in an area, a mode that does not wouldn't contribute that much anyway.) Here, it so happens that from what I saw, several papers have excellent visualizations[^4], and the same holds for some lecture slides and accompanying notebooks[^5]. But for those among us with limited spatial-imagination capabilities -- e.g., people with [Aphantasia](https://en.wikipedia.org/wiki/Aphantasia) -- these illustrations, intended to help, can be very hard to make sense of themselves. If you're not one of these, I totally recommend checking out the resources linked in the above footnotes. This text, though, will try to make the best possible use of verbal explanation to introduce the concepts involved, the library, and how to use it.
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:44]: > This text, though, will try to make the best possible use of verbal explanation to introduce the concepts involved, the library, and how to use it.

OK, good. Now I know what you're about to be telling me! There are a lot of caveats and asides in these first few paragraphs, so it's _very_ helpful to end with a clear declarative statement at the end.

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+
+[^4]: One paper particularly stood out to me: [@abs-2106-06020].
+
+[^5]: Directly pertinent to today's topic, thinking of the materials produced for University of Amsterdam's [course on group-equivariant deep learning](https://uvagedl.github.io/).
+
+That said, let's start with the software.
+
+## Using *escnn*
+
+`Escnn` depends on PyTorch. Yes, PyTorch, not `torch`; unfortunately, the library hasn't been ported to R yet. (If, after reading this post, you feel that maybe *you* would be interested in porting it -- I can definitely say I think that's a great idea!) For now, thus, we'll employ [`reticulate`](https://rstudio.github.io/reticulate/)[^6] to access the Python objects directly.
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:56]: > (If, after reading this post, you feel that maybe *you* would be interested in porting it -- I can definitely say I think that's a great idea!)

This could maybe be a footnote rather than a parenthetical, since you're starting to really jump into the explanation now.

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+py_list_attributes(gspaces) |> (\(vec) grep("On", vec, value = TRUE))() |> sort()
+```
+
+```         
+[1] "conicalOnR3" "cylindricalOnR3" "dihedralOnR3" "flip2dOnR2" "flipRot2dOnR2" "flipRot3dOnR3"
+[7] "fullCylindricalOnR3" "fullIcoOnR3" "fullOctaOnR3" "icoOnR3" "invOnR3" "mirOnR3 "octaOnR3"
+[14] "rot2dOnR2" "rot2dOnR3" "rot3dOnR3" "trivialOnR2" "trivialOnR3"    
+```
+
+The methods I've listed instantiate a `gspace`. If you look closely, you see that they're all composed of two strings, joined by "On". In all instances, the second part is either `R2` or `R3`. These two are the available base spaces -- $\mathbb{R}^2$ and $\mathbb{R}^3$ -- an input signal can live in. Signals can, thus, be images, made up of pixels, or three-dimensional volumes, composed of voxels. The first part refers to the group you'd like to use. Choosing a group means choosing the symmetries to be respected. For example, `rot2dOnR2()` implies equivariance as to rotations, `flip2dOnR2()` guarantees the same for mirroring actions, and `flipRot2dOnR2()` subsumes both.
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:93]: Great paragraph. Very clear.

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+$irrep_2
+C4|[irrep_2]:1
+
+$regular
+C4|[regular]:4
+```
+
+A representation, in our current context *and* very roughly speaking, is a way to encode a group action as a matrix, meeting certain conditions. In `escnn`, representations are central, and we'll see how in the next section.
+
+First, let's inspect the above output. Four representations are available, three of which share an important property: they're all [irreducible](https://en.wikipedia.org/wiki/Irreducible_representation). On $C_4$, any non-irreducible representation [can be decomposed](https://en.wikipedia.org/wiki/Maschke%27s_theorem) into into irreducible ones. These irreducible representations are what `escnn` works with internally. Of those three, the most interesting one is the second. To see its action, we need to choose a group element. How about counterclockwise rotation by ninety degrees:
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:135]: Again, really good stuff—very clear even to someone who isn't familiar with the vocabulary.

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+
+This is a step-one shift to the right of the identity matrix. The identity matrix, mapped to element 0, is the non-action; this matrix instead maps the zeroth action to the first, the first to the second, the second to the third, and the third to the first.
+
+We'll see the regular representation used in a neural network soon. Internally -- but that need not concern the user -- *escnn* works with its decomposition into irreducible matrices. Here, that's just the bunch of irreducible representations we saw above, numbered from one to three.
+
+Having looked at how groups and representations figure in `escnn`, it is time we approach the task of building a network.
+
+## Representations, for real: `escnn$nn$FieldType`
+
+So far, we've characterized the input space ($\mathbb{R}^2$), and specified the group action. But once we enter the network, we're not in the plane anymore, but in a space that has been extended by the group action. Rephrasing, the group action produces *feature vector fields* that assign a feature vector to each spatial position in the image.
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:203]: > So far, we've characterized the input space ($\mathbb{R}^2$), and specified the group action.

Good summary.

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+
+Now we have those feature vectors, we need to specify how they transform under the group action. This is encoded in an `escnn$nn$FieldType` . Informally, we could say that a field type is the *data type* of a feature space. In defining it, we indicate two things: the base space, a `gspace`, and the representation type(s) to be used.
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:205]: > Now we have those feature vectors, we need to specify how they transform under the group action.

Again, this is really helpful and clear! _This_ is 100% the type of writing I'm referring to when I talk about helping a reader to avoid getting lost! 🙂

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 11]
```diff
@@ -0,0 +1,492 @@
...
+You see that at the end, we are back at the original "image".
+
+Now, for equivariance. We could first apply a rotation, then convolve. Rotate:
+
+```{r}
+x_rot <- x$transform(g1)
+x_rot$tensor
+```
+
+This is the first in the above list of four tensors. -- Convolve:
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:331]: I'm not sure what the relationship between the words/sentences is meant to be when you have a period followed by an em dash and a word with a capital letter, e.g.,
> . -- Convolve
I'm not suggesting anything else, because I don't know what you're going for.

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 12]
```diff
@@ -0,0 +1,492 @@
...
+          [-0.3743,  2.8144]],
+
+         [[ 5.0640, 11.7395],
+          [ 8.6488, 31.7169]],
+
+         [[ 2.3499,  1.7937],
+          [ 4.5065,  5.9689]]]], grad_fn=<ConvolutionBackward0>)
+```
+
+Alternatively, we can do the convolution first, then rotate its output. Convolve:
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:352]: OK, here you have it without the em dash, so maybe do the same above.

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 13]
```diff
@@ -0,0 +1,492 @@
...
+         [[ 5.0640, 11.7395],
+          [ 8.6488, 31.7169]],
+
+         [[ 2.3499,  1.7937],
+          [ 4.5065,  5.9689]]]])
+```
+
+Indeed, final results are the same.
+
+At this point, we know how to employ group-equivariant convolutions. The final step is to compose the network.
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:396]: Again, great transition.

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+
+## A group-equivariant neural network
+
+Basically, we have two questions to answer. The first concerns the non-linearities. This is a potentially intricate topic, but as long as we stay with point-wise operations (such as that performed by ReLU) equivariance is given intrinsically.
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:400]: > we have two questions to answer.

I think you delve deeply into the first before you get to the second, so you might want to either say what they both are now, or start this with something to the effect of “The first question we have to answer concerns the non-linearities."

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd [hunk 15]
```diff
@@ -0,0 +1,492 @@
...
+
+## Where to from here
+
+This "introduction to an introduction" has been the attempt to draw a high-level map of the terrain, so you can decide if this is useful to you. If it's not just useful, but interesting theory-wise, as well, you'll find lots of excellent materials linked from the [README](https://github.com/QUVA-Lab/escnn). The way I see it, though, this post already should enable you to actually experiment with different setups.
+
+One such experiment, that would be of high interest to me, might investigate how well different types and degrees of equivariance actually work for a given task and dataset. Overall, a reasonable assumption is that, the "up" we go in the feature hierarchy, the less equivariance we require. For edges and corners, taken by themselves, full rotation equivariance seems desirable, as does equivariance to reflection; for higher-level features, we might want to successively restrict allowed operations, maybe ending up with equivariance to mirroring merely. Experiments could be designed to compare different ways, and levels, of restriction.[^7]
```
Comment (batpigandme) [inline _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:484]:
```suggestion
One such experiment, that would be of high interest to me, might investigate how well different types and degrees of equivariance actually work for a given task and dataset. Overall, a reasonable assumption is that, the higher "up" we go in the feature hierarchy, the less equivariance we require. For edges and corners, taken by themselves, full rotation equivariance seems desirable, as does equivariance to reflection; for higher-level features, we might want to successively restrict allowed operations, maybe ending up with equivariance to mirroring merely. Experiments could be designed to compare different ways, and levels, of restriction.[^7]
```

Comment (batpigandme) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:484]: Or, if you don't want to use higher/lower, you can use more/less, increase/decreate, etc.—it's just that "the up we go…" doesn't read correctly.

**Review by batpigandme — APPROVED** [review]
The code-filled sections are especially well done. I've added notes in some areas where I might have gotten a bit lost, and they're almost exclusively in the parts where you're sort of "meta" describing the post. 

Remember, _you_ know more than your reader does about what you're _not_ including. You don't need to apologize for it! It's helpful when you point out different resources, but, it can be confusing if there are so many asides/caveats that the reader can't remember what you ARE about to tell us!

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd (hunk 14 above)
Comment (batpigandme) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:400]: Having read the whole section, I'm not totally sure I get what that second question is 😬

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd (hunk 1 above)
Comment (skeydan) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:28]: Interesting! Will keep that in mind. Here rules between languages definitely are different (in German, for instance, you'd capitalize when what follows is a complete sentence (here, there'd be several)

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd (hunk 4 above)
Comment (skeydan) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:40]: I meant it in sort of a more general way... Certainly, what's included are the library objects (B), but a broader understanding where the math terms (A), too, are characters of the story, works fine as well...

I think that overall in this post, I'm playing "fast and loose", so to say, with the writing style, trying to combine maximal precision and entertainment (insofar that is possible with such a topic)...

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd (hunk 5 above)
Comment (skeydan) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:44]: Yeah, you're right :-) I know, there's many caveats & all ... I also know I usually don't go into details about what makes things hard, but in this post, I really wanted to, subjective and personal though it may be...

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd (hunk 3 above)
Comment (skeydan) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:38]: Yes, exactly

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd (hunk 13 above)
Comment (skeydan) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:396]: thank you!

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd (hunk 6 above)
Comment (skeydan) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:56]: Oh, true, will do! Too much of a "sales pitch" (or the opposite, sort of)

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd (hunk 7 above)
Comment (skeydan) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:93]: thanks!

Comment (skeydan) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:93]: thanks!

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd (hunk 10 above)
Comment (skeydan) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:205]: I'm very happy to hear that, thank you!

**Review by dfalbel — APPROVED** [review]
Looks good @skeydan ! I don't have additional comments.

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd (hunk 2 above)
Comment (skeydan) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:30]: Yes, I see what you mean. I was thinking about having separate paragraphs, but then deciding against it, for the usual bad reasons ... (_doesn't seem important enough_)

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd (hunk 8 above)
Comment (skeydan) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:135]: thanks!!

File: _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd (hunk 14 above)
Comment (skeydan) [reply _posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd:400]: Oh yes that's correct! I'll do as you say, and name the second immediately.

**Comment by skeydan** [comment]
Thanks a lot @batpigandme & @dfalbel !
@batpigandme it is just incredible how you infallibly spot any deficiencies in the reasoning, whatever the topic! That's really quite miraculous :-)
